### DISCUSSION TITLE
Low-level interface CO2 transcritical phase problem

### DISCUSSION THREAD

**Discussion by AliSalami98** [CoolProp/CoolProp#2438] [Q&A]
Hello all,

I've been using low-level interface CoolProp since two years now in a python environment. Previously I had errors when CO2 pressure is between 73.7 and 73.8 (around the critical pressure) so I had to avoid this 0.1 bar range by forcing one of the boundary values. But no after updating CoolProp I have this error: 

ValueError: unable to solve 1phase PY flash with Tmin=303.874, Tmax=3000 due to error: solver_rho_Tp was unable to find a solution for T=   303.874, p=7.33593e+06, with guess value    7905.78 with error: Householder4 reached maximum number of iterations

Any ideas what might be the problem?

**Comment by ibell** [comment]
Can you please provide an input or two demonstrating the problem? @msaitta-mpr has done nice work on updating iterative calculations and might be able to dig in.

**Reply by msaitta-mpr** [reply]
Interesting. I am able to replicate the issue with:
```Python
import CoolProp as cp
state = cp.AbstractState("HEOS", "CO2")
p = 7377262.928140703
u = 314719.5306503257
state.update(cp.PUmass_INPUTS, p, u)
```

I will try to dig in and figure out what is happening here.

**Comment by AliSalami98** [comment]
Thanks. I appreciate it.

**Comment by msaitta-mpr** [comment]
I spent some time digging into this and the issue has to do with a bad phase determination near the critical point. I need a little more time to determine how to fix the issue.

**Comment by AliSalami98** [comment]
Hello,

Sorry to bother you again with this. But I think the error is still there.

Ali

**Comment by ibell** [comment]
Have you tried with version 7?

**Comment by AliSalami98** [comment]
Yes. The error is still there.

Le jeu. 14 août 2025 à 12:47 AM, Ian Bell ***@***.***> a
écrit :

**Comment by ibell** [comment]
Thanks for letting me know. I _think_ I know how to fix this,

**Comment by AliSalami98** [comment]
Hello Ian,

I believe that version 7 is 4 to 5 times slower in execution. I just tried
this command with version 7 and version 6.8: 
```
from CoolProp.CoolProp import
AbstractState
import CoolProp.CoolProp as CP
import time

def get_state(input_type, input1, input2):
    state = AbstractState("HEOS", "CO2")
    state.update(input_type, input1, input2)
    return state

pc = 70e5
hc = 300e3
N = 10_000

# warm-up (jit/cache, DLL load, etc.)
get_state(CP.HmassP_INPUTS, hc, pc)

t0 = time.perf_counter()
for _ in range(N):
    state = get_state(CP.HmassP_INPUTS, hc, pc)
t1 = time.perf_counter()

dt = t1 - t0
print(f"Total time: {dt:.4f} s")
print(f"Avg per call: {1e6*dt/N:.1f} µs")
print(f"Calls per second: {N/dt:.0f} 1/s")
```

**Comment by AliSalami98** [comment]
Sorry but I'm a daily user of Coolprop and use it in my dynamic simulation
so I'll notice speed differences. I tried the code I sent you with both
versions:

- 6.8 took 4.5 sec to execute while version 7 took 26 sec

**Comment by ibell** [comment]
Ok there is definitely a performance regression in this case. I will look into it; it is relatedd to the superancillary functions somehow, it would seem. Docs here: https://coolprop.org/coolprop/SuperAncillary.html

I extended your testing script to totally disable the superancillary functions (not even load them), and that made a huge difference to the speed, and I have literally no idea why. Script:
```python
import os 
os.environ['COOLPROP_DISABLE_SUPERANCILLARIES_ENTIRELY'] = '1'

from CoolProp.CoolProp import AbstractState
import CoolProp.CoolProp as CP
import time

def get_state(input_type, input1, input2):
    state = AbstractState("HEOS", "CO2")
    return state

def get_state_and_update(input_type, input1, input2):
    state = AbstractState("HEOS", "CO2")
    state.update(input_type, input1, input2)
    return state

def do_update(state, input_type, input1, input2):
    state.update(input_type, input1, input2)
    return state

pc = 70e5
hc = 300e3
N = 10_000

# warm-up (jit/cache, DLL load, etc.)
get_state(CP.HmassP_INPUTS, hc, pc)

print("Benchmarking get_state")

t0 = time.perf_counter()
for _ in range(N):
    state = get_state(CP.HmassP_INPUTS, hc, pc)
t1 = time.perf_counter()

dt = t1 - t0
print(f"Total time: {dt:.4f} s")
print(f"Avg per call: {1e6*dt/N:.1f} µs")
print(f"Calls per second: {N/dt:.0f} 1/s")


print("Benchmarking get_state_and_update")

t0 = time.perf_counter()
for _ in range(N):
    state = get_state_and_update(CP.HmassP_INPUTS, hc, pc)
t1 = time.perf_counter()

dt = t1 - t0
print(f"Total time: {dt:.4f} s")
print(f"Avg per call: {1e6*dt/N:.1f} µs")
print(f"Calls per second: {N/dt:.0f} 1/s")


print("Benchmarking do_update")

t0 = time.perf_counter()
for _ in range(N):
    state = do_update(state, CP.HmassP_INPUTS, hc, pc)
t1 = time.perf_counter()

dt = t1 - t0
print(f"Total time: {dt:.4f} s")
print(f"Avg per call: {1e6*dt/N:.1f} µs")
print(f"Calls per second: {N/dt:.0f} 1/s")
```

With superancillaries off entirely:
```
CoolProp: superancillaries have been disabled because the COOLPROP_DISABLE_SUPERANCILLARIES_ENTIRELY environment variable has been defined
Benchmarking get_state
Total time: 0.5824 s
Avg per call: 58.2 µs
Calls per second: 17172 1/s
Benchmarking get_state_and_update
Total time: 3.4391 s
Avg per call: 343.9 µs
Calls per second: 2908 1/s
Benchmarking do_update
Total time: 2.9084 s
Avg per call: 290.8 µs
Calls per second: 3438 1/s
```

and with the first two lines commented out:
```
Benchmarking get_state
Total time: 2.0899 s
Avg per call: 209.0 µs
Calls per second: 4785 1/s
Benchmarking get_state_and_update
Total time: 17.7555 s
Avg per call: 1775.5 µs
Calls per second: 563 1/s
Benchmarking do_update
Total time: 15.5926 s
Avg per call: 1559.3 µs
Calls per second: 641 1/s
```

**Comment by AliSalami98** [comment]
Thanks a lot Ian! This is very helpful!

Le ven. 15 août 2025 à 12:39 AM, Ian Bell ***@***.***> a
écrit :

**Comment by ibell** [comment]
I have fixed that particular issue, now it is hundreds of times faster than 6.8.0! Changes landing on main in a few seconds. I am going to do the same for T, {D,S} this weekend. I am sure the same problem is present there. It comes down to making copies of state classes internally.

**Comment by AliSalami98** [comment]
Highly appreciated! No excuse for using lookup-tables :-)

And about the problem around the critical point?

Le ven. 15 août 2025 à 1:12 AM, Ian Bell ***@***.***> a
écrit :

**Comment by ibell** [comment]
I will get around to that, hopefully soon. But for now, I need to focus on the very serious performance regressions: https://github.com/CoolProp/CoolProp/issues/2592
